File: jscomp/test/gpr_2642_test.js
'use strict';


function isfree(id, _id$p) {
  while(true) {
    var id$p = _id$p;
    switch (id$p.TAG) {
      case "Pident" :
          return id === id$p._0;
      case "Pdot" :
          _id$p = id$p._0;
          continue ;
      case "Papply" :
          if (isfree(id, id$p._0)) {
            return true;
          }
          _id$p = id$p._1;
          continue ;
      
    }
  };
}

exports.isfree = isfree;
/* No side effect */

File: jscomp/test/gpr_2642_test.ml
type t =
  Pident of string
| Pdot of t * string * int
| Papply of t * t


let rec isfree id = function
Pident id' ->  id = id'
| Pdot(p, s, pos) ->
  isfree id p 
  (* if isfree id p then
    true
  else false *)
| Papply(p1, p2) -> isfree id p1 || isfree id p2
